components/sendgrid/sendgrid.app.js

```diff
@@ -118,5 +113,429 @@ module.exports = {
+     *
+     * @param {integer}  startTime start of the time range in unix timestamp when a block was
+     * created (inclusive).
+     * @param {integer}  endTime end of the time range in unix timestamp when a block was
+     * created (inclusive).
+     * @param {integer}  numberOfBlocks the number of blocks to return.
+     * @returns {{created: number, email: string, reason: string, status: string}: array} an array
+     * with details of each block returned: `created` for unix timestamp for when the block record
+     * was created at SendGrid, `email` for the email address that was added to the block  list,
+     * `reason` with the reason for the block, and the `status` of the block.
+     */
+    async listBlocks(startTime, endTime, maxItems) {
+      const pageSize = Math.min(maxItems, 100);
+      const blocks = [];
+      let url = `/v3/suppression/blocks?limit=${pageSize}`;
+      let config = {
+        method: "GET",
+        url,
+        start_time: startTime,
+        end_time: endTime,
+      };
+      let lastIteration = false;
+      do {
+        const data  = (await this._withRetries(() => this.api().request(config)));
+        blocks.push(...data[1]);
+        if (lastIteration) {
+          break;
+        }
+        const links = data[0].headers.link.split(",");
+        url = "";
+        links.forEach( (link) => {
+          let next = "";
+          let last = "";
+          if ( link.indexOf("next") > -1) {
+            const idx = link.indexOf(";");
+            next = link.substring(0, idx).replace("<", "")
+              .replace(">", "");
+            url = next;
+          }
+          if ( link.indexOf("last") > -1) {
+            const idx = link.indexOf(";");
+            last = link.substring(0, idx).replace("<", "")
+              .replace(">", "");
+          }
+          const nextArr = [];
+          nextArr.push(next);
+          if (nextArr.includes(last)) {
+            lastIteration = true;
+          }
+        });
+        config.url = url;
+      } while (blocks.length < maxItems);
+      return blocks.slice(0, maxItems);
+    },
+    /**
+     * Lists all email addresses that are globally suppressed.
+     *
+     * @param {integer}  startTime start of the time range in unix timestamp when a global
+     * supression was created (inclusive).
+     * @param {integer}  endTime end of the time range in unix timestamp when  a global
+     * supression was created (inclusive).
+     * @param {integer}  numberOfSupressions the number of global supressions to return.
+     * @returns {{created: number, email: string}: array} an array
+     * with details of each global supression returned: `created` for unix timestamp for when
+     * the recipient was added to the global suppression list, `email` for the email address of
+     * the recipient who is globally suppressed.
+     */
+    async listGlobalSupressions(startTime, endTime, maxItems) {
```

Comment:
It seems like this function along with `listBlocks` function have similar patterns. How about abstracting this and put them in a single function call?

Reply:
Yeah that should do. Abstracting out into a single function call was easy.